yarpl/include/yarpl/flowable/EmitterFlowable.h

```diff
@@ -26,25 +26,20 @@ class EmiterBase : public virtual Refcounted {
  * This is synchronous: the emit calls are triggered within the context
  * of a request(n) call.
  */
-template<typename T>
-class EmiterSubscription : private Subscription, private Subscriber<T> {
+template <typename T>
+class EmiterSubscription : public virtual Subscription,
+                           public virtual Subscriber<T> {
   constexpr static auto kCanceled = credits::kCanceled;
   constexpr static auto kNoFlowControl = credits::kNoFlowControl;
 
  public:
   EmiterSubscription(
-      Reference <EmiterBase<T>> emiter,
-      Reference <Subscriber<T>> subscriber)
-      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {
-    // We expect to be heap-allocated; until this subscription finishes
-    // (is canceled; completes; error's out), hold a reference so we are
-    // not deallocated (by the subscriber).
-    Refcounted::incRef(*this);
-    subscriber_->onSubscribe(get_ref<Subscription>(this));
-  }
+      Reference<EmiterBase<T>> emiter,
+      Reference<Subscriber<T>> subscriber)
+      : emiter_(std::move(emiter)), subscriber_(std::move(subscriber)) {}
 
-  virtual ~EmiterSubscription() {
-    subscriber_.reset();
+  void init() {
+    subscriber_->onSubscribe(get_ref<Subscription>(this));
```

Comment:
It looks like either we increment count in the constructor and decrement it in the release function, or we first create an instance of Reference<T> and later pass this instance to another object. Passing an instance of itself, get_ref(this), in the constructor and passing it around causes the object to get deleted. To prevent this they were calling inc_ref.

This is an improvement on that at least and I believe an improvement is good to check in :)

https://github.com/rsocket/rsocket-cpp/issues/671